### DISCUSSION TITLE
Inconsistent behavior

### DISCUSSION THREAD

**Discussion by 0x7FFFFFFFFFFFFFFF** [rust-shell-script/rust_cmd_lib#37] [General]
I'm using version 1.2.2 and I have the following program.

```
fn main() -> eyre::Result<()> {
    std::env::set_current_dir(r##"f:\temp"##);
    let output = cmd_lib::spawn_with_output!(ls -la)?.wait_fun_result()?;

    println!("{:?}", output);
    Ok(())
}
```

When I run it, I can get the expected result like:
```
"total 5996106\ndrwxr-xr-x 1 admin None         0 Sep 25 16:26 .\n..............................."
```

When I change the command `ls -la` to more complex one (which probably takes several seconds to complete), I get empty output. 

```
fn main() -> eyre::Result<()> {
    std::env::set_current_dir(r##"f:\temp"##);
    let output = cmd_lib::spawn_with_output!(
        ffmpeg -i 1.mp4 -i a2.png -y -filter_complex "overlay=x=263:y=752"  -c:a copy -max_muxing_queue_size 99999 -y test.mp4
    )?.wait_fun_result()?;

    println!("{:?}", output);
    Ok(())
}
```
Output is:
```
""
```

But can I find the produced output file of the ffmpeg command, which indicates the command did run. If I enabled the logger like this:

```
use cmd_lib::init_builtin_logger;
fn main() -> eyre::Result<()> {
    init_builtin_logger();
    std::env::set_current_dir(r##"f:\temp"##);
    let output = cmd_lib::spawn_with_output!(
        ffmpeg -i 1.mp4 -i a2.png -ss 00:00:00 -to 00:01:00 -y -filter_complex "overlay=x=263:y=752"  -c:a copy -max_muxing_queue_size 99999 -y testtttt.mp4
    )?.wait_fun_result()?;
    println!("{:?}", output);
    Ok(())
}
```

I can get the ffmpeg output in the console:
```
INFO - ffmpeg version N-103679-g7bbad32d5a-20210918 Copyright (c) 2000-2021 the FFmpeg developers
INFO -   built with gcc 10-win32 (GCC) 20210408
INFO -   configuration: --prefix=/ffbuild/prefix --pkg-config-flags=--static --pkg-config=pkg-config --cross-prefix=x86_64-w64-
...
INFO - [libx264 @ 0000010cf15da6c0] ref B L1: 97.3%  2.7%
INFO - [libx264 @ 0000010cf15da6c0] kb/s:134.12
""
```

But note the `""` at the end of the output, I still can't get the output in the variable `output` and it still contains an empty string. 

Why for simple command, the output is captured in a variable. But for complex command, it doesn't do this. Is this inconsistence a bug? How can I store the command output in a variable when the command is kinda complex? Thank.

**Comment by tao-guo** [comment]
By default only stdout output is captured into variable, and stderr is logged into your logger. You can redirect stderr to stdout if you want to capture  it into variable: `let v = run_fun!(xxx 2>&1)?;`. Also don't use spawn API if you are not doing things concurrently.

**Comment by 0x7FFFFFFFFFFFFFFF** [comment]
Thanks for the suggestion. I changed the command to this:
```
let output = cmd_lib::spawn_with_output!(
        ffmpeg -i 1.mp4 -i a2.png -ss 00:00:00 -to 10 -y -filter_complex "overlay=x=263:y=752"  -c:a copy -max_muxing_queue_size 99999 -y testtttt.mp4 2>result.txt
    )?.wait_fun_result()?;
```
Since ffmpeg always log to stderr, and I redirected stderr to result.txt, I can get the log in the text file and the `output` variable is empty. However, when I change the command to this:
```
let output = cmd_lib::spawn_with_output!(
        ffmpeg -i 1.mp4 -i a2.png -ss 00:00:00 -to 10 -y -filter_complex "overlay=x=263:y=752"  -c:a copy -max_muxing_queue_size 99999 -y testtttt.mp4 2>&1
    )?.wait_fun_result()?;
```
Here I redirect stderr to stdout, which should make the variable `output` capture both stdout and stderr. However, the behavior is that the whole output get's logged in the terminal and the variable `output` is still empty. Where am I wrong?

**Comment by rust-shell-script** [comment]
Try `let output = run_fun!(ffmgeg ... |& cat)?` to see if it works. The Stdio::piped() API has some known issues for redirection, and it will be replaced in the future when I have enough time to fix it.

**Reply by 0x7FFFFFFFFFFFFFFF** [reply]
> Try `let output = run_fun!(ffmgeg ... |& cat)?` to see if it works. The Stdio::piped() API has some known issues for redirection, and it will be replaced in the future when I have enough time to fix it.

Yes, this works! Thanks for the trick. Do you want me to open an issue to track the redirection improvement you mentioned?
